### ISSUE TITLE
Severe off-target issues when reproducing Many2Many translation

### ISSUE THREAD

**Issue by Smu-Tan** [voidmagic/parameter-differentiation#2]
Hi! Thank you for releasing the implementation!

I'm trying to reproduce your code on the iwslt data using the template you provided, however, I got severe off-target issues when generating en-xx translations. 

[Problem I got] I found when generating en-xx, the actual generations are always in English, even though the langtoks are specified.

I would appreciate it if could you please provide more details for reproducing Many2Many translations. 

Below I list my training and inference code, thank you in advance!

## Training
```
#!/bin/bash
DATA_DIR=iwslt14/iwslt14-30k/data_bin
CHECKPOINT_DIR=checkpoints/iwslt14
fairseq-train ${DATA_DIR} \
    --user-dir . \
    --save-dir ${CHECKPOINT_DIR} \
    --lang-pairs en-ar,ar-en,en-de,de-en,en-es,es-en,en-fa,fa-en,en-he,he-en,en-it,it-en,en-nl,nl-en,en-pl,pl-en \
    --arch parameter_differentiation_iwslt_model \
    --task parameter_differentiation_task \
    --encoder-layers 3 --decoder-layers 3 \
    --share-all-embeddings --fp16 \
    --encoder-langtok src --decoder-langtok \
    --criterion label_smoothed_cross_entropy \
    --label-smoothing 0.1 --dropout 0.1 --skip-invalid-size-inputs-valid-test \
    --optimizer adam --adam-betas '(0.9, 0.98)' \
    --clip-norm 0.0 --lr 5e-4 --lr-scheduler inverse_sqrt --warmup-updates 4000 --weight-decay 0.0 \
    --max-tokens 8192 --update-freq 5 --patience 20 --max-update 200000 \
    --no-epoch-checkpoints --seed 222 \
    --validate-interval 1
```

## inference
```
PAIRS=('en-ar' 'en-de' 'en-es' 'en-fa' 'en-he' 'en-it' 'en-nl' 'en-pl' 'ar-en' 'de-en' 'es-en' 'fa-en' 'he-en' 'it-en' 'nl-en' 'pl-en')
for i in "${!PAIRS[@]}"; do
    PAIR=${PAIRS[i]}
    SRC=${PAIR%-*}
    TGT=${PAIR#*-}
    SAVE_DIR=$CHECKPOINT_DIR/${SRC}-${TGT}
    fairseq-generate $DATA_DIR \
        --lang-pairs en-ar,ar-en,en-de,de-en,en-es,es-en,en-fa,fa-en,en-he,he-en,en-it,it-en,en-nl,nl-en,en-pl,pl-en \
        --source-lang $SRC \
        --target-lang $TGT \
        --user-dir . --max-tokens 30000 \
        --task parameter_differentiation_task \
        --arch parameter_differentiation_iwslt_model \
        --remove-bpe sentencepiece \
        --path ${CHECKPOINT_DIR}/checkpoint_best.pt \
        --skip-invalid-size-inputs-valid-test \
        --encoder-langtok src --decoder-langtok \
        --results-path $SAVE_DIR --beam 5
done
```

**Comment by voidmagic** [comment]
Have you tried the vanilla MNMT and does it work?  I think the most possible reason for off target issue is the language token. Please make sure the token is consistent  during training and inference. BTW I have never seen off target issue, even when the training is incorrect (the translation quality is low but the language is still correct)

**Comment by Smu-Tan** [comment]
> Have you tried the vanilla MNMT and does it work? I think the most possible reason for off target issue is the language token. Please make sure the token is consistent during training and inference. BTW I have never seen off target issue, even when the training is incorrect (the translation quality is low but the language is still correct)

Thank you for the reply! Yes, I have trained the vanilla MMT with `multilingual_translation` and it does not lead to off-target issues. I also checked the tags and they are well prepended, I will double-check again!

**Comment by Smu-Tan** [comment]
When setting the fairseq and pytorch versions to the recommended ones, the off-target issue is solved! BTW, I found the algorithm cannot be run with multi-GPU training, could you let me know if is this a bug or if it simply does not support multi-GPU training? Thank you in advance!

**Comment by voidmagic** [comment]
> When setting the fairseq and pytorch versions to the recommended ones, the off-target issue is solved! BTW, I found the algorithm cannot be run with multi-GPU training, could you let me know if is this a bug or if it simply does not support multi-GPU training? Thank you in advance!

It does not support multi gpu training and we use single gpu + gradient accumulation in our experiments. To use multi gpu, the model structure should be synced across devices all the time, which requires something like StateManager between view and trainer.

**Comment by Smu-Tan** [comment]
> > When setting the fairseq and pytorch versions to the recommended ones, the off-target issue is solved! BTW, I found the algorithm cannot be run with multi-GPU training, could you let me know if is this a bug or if it simply does not support multi-GPU training? Thank you in advance!
> 
> It does not support multi gpu training and we use single gpu + gradient accumulation in our experiments. To use multi gpu, the model structure should be synced across devices all the time, which requires something like StateManager between view and trainer.

Thank you for the reply! Could you please guide how to set the "differentiation upper bound" for the current implementation?

**Comment by Smu-Tan** [comment]
I implemented it by myself, for people who are also interested in it, see the code below:

```
def begin_valid_epoch(self, epoch, model):
        # record the initial number of params
        if epoch==1:
            self.initial_num_param=sum(p.numel() for p in model.parameters())
            print('\nThe Initial Number of Params is: ', self.initial_num_param)

        self.record_gradient(model)
        
        # determine if the param ratio exceeds the threshold
        cur_num_param=sum(p.numel() for p in model.parameters())
        if cur_num_param/self.initial_num_param <= self.args.upper_bound:
            logger.info("num. model params before: {}".format(sum(p.numel() for p in model.parameters())))
            _ = list(self.view.auto_split())
            logger.info("num. model params after: {}".format(sum(p.numel() for p in model.parameters())))
        else:
            print('param ratio is: ', cur_num_param/self.initial_num_param)
            print('Now stop expanding models')

        self.view.clear_gradient()
        get_trainer()._optimizer = None
```
